File: backend/services/pms_service/app/schemas/sales.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from .inventory_validation import Money, Units
from .transactions import PaymentInput, TransactionInput


class WalkInSaleItemIn(TransactionInput):
    drug_id: UUID
    quantity: Units
    # Optional override; default is the batch's selling_price_cents.
    unit_price_cents: Money | None = None


class WalkInSaleCreate(PaymentInput):
    customer_id: UUID | None = None
    items: list[WalkInSaleItemIn] = Field(min_length=1, max_length=100)
    discount_cents: Money = 0
    tax_cents: Money = 0

    @field_validator("items")
    @classmethod
    def unique_drugs(cls, items):
        if len({item.drug_id for item in items}) != len(items):
            raise ValueError("include each drug only once")
        return items


class SaleItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    sale_id: UUID
    prescription_item_id: UUID | None = None
    corrected_quantity: int = 0
    drug_id: UUID
    drug_batch_id: UUID
    drug_name_snapshot: str
    quantity: int
    unit_price_cents: int
    line_total_cents: int


class SaleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    sale_number: str
    version: int
    notes: str | None
    void_reason: str | None
    prescription_id: UUID | None
    customer_id: UUID | None
    cashier_staff_id: UUID | None
    subtotal_cents: int
    discount_cents: int
    tax_cents: int
    total_cents: int
    currency: str
    payment_method: str
    payment_ref: str | None
    status: str
    completed_at: datetime | None
    created_at: datetime
    items: list[SaleItemOut] = []
    credited_cents: int = 0
    refunded_cents: int = 0
    refundable_cents: int = 0


class SaleList(BaseModel):
    items: list[SaleOut]
    total: int
    limit: int
    offset: int

File: frontend/pms_web/src/lib/repositories/sales.ts
import apiClient from "@/lib/api/client";
import type {
  CancelTransaction,
  Page,
  PaymentDetails,
  TransactionQuote,
} from "./transactions";

export interface SaleItem {
  id: string;
  sale_id: string;
  prescription_item_id?: string | null;
  corrected_quantity?: number;
  drug_id: string;
  drug_batch_id: string;
  drug_name_snapshot: string;
  quantity: number;
  unit_price_cents: number;
  line_total_cents: number;
}

export interface Sale {
  id: string;
  sale_number: string;
  version: number;
  notes?: string | null;
  void_reason?: string | null;
  prescription_id?: string | null;
  customer_id?: string | null;
  cashier_staff_id?: string | null;
  subtotal_cents: number;
  discount_cents: number;
  tax_cents: number;
  total_cents: number;
  currency: string;
  payment_method: string;
  payment_ref?: string | null;
  status: "completed" | "voided";
  completed_at?: string | null;
  created_at: string;
  items: SaleItem[];
  credited_cents?: number;
  refunded_cents?: number;
  refundable_cents?: number;
}

export interface WalkInSaleCreate extends PaymentDetails {
  customer_id?: string;
  items: { drug_id: string; quantity: number; unit_price_cents?: number }[];
  discount_cents?: number;
  tax_cents?: number;
}

export const salesRepo = {
  page: async (
    params: {
      status?: string;
      payment_method?: string;
      prescription_id?: string;
      search?: string;
      offset?: number;
      limit?: number;
    },
    signal?: AbortSignal,
  ): Promise<Page<Sale>> => {
    const { data } = await apiClient.get("/v1/sales", { params, signal });
    return data;
  },
  get: async (id: string, signal?: AbortSignal): Promise<Sale> => {
    const { data } = await apiClient.get(`/v1/sales/${id}`, { signal });
    return data;
  },
  quote: async (
    body: WalkInSaleCreate,
    signal?: AbortSignal,
  ): Promise<TransactionQuote> => {
    const { data } = await apiClient.post("/v1/sales/quote", body, { signal });
    return data;
  },
  createWalkIn: async (
    body: WalkInSaleCreate,
    requestId: string,
    signal?: AbortSignal,
  ): Promise<Sale> => {
    const { data } = await apiClient.post("/v1/sales", body, {
      signal,
      headers: { "Idempotency-Key": requestId },
    });
    return data;
  },
  voidSale: async (
    id: string,
    body: CancelTransaction,
    requestId: string,
    signal?: AbortSignal,
  ): Promise<Sale> => {
    const { data } = await apiClient.post(`/v1/sales/${id}/void`, body, {
      signal,
      headers: { "Idempotency-Key": requestId },
    });
    return data;
  },
};
